kernel/src/process_checking.rs

```diff
@@ -0,0 +1,356 @@
+//! Traits, types, and sample implementations of application credentials
+//! checkers, used to decide whether an application can be loaded. See
+//| the [AppID TRD](../../doc/reference/trd-appid.md).
+
+use crate::dynamic_deferred_call::{
+    DeferredCallHandle, DynamicDeferredCall, DynamicDeferredCallClient,
+};
+use crate::hil::digest::{ClientData, ClientHash, ClientVerify};
+use crate::hil::digest::{DigestDataVerify, Sha256};
+use crate::process::{Process, ShortID, State};
+use crate::utilities::cells::OptionalCell;
+use crate::utilities::cells::TakeCell;
+use crate::utilities::leasable_buffer::{LeasableBuffer, LeasableMutableBuffer};
+use crate::ErrorCode;
+use tock_tbf::types::TbfFooterV2Credentials;
+use tock_tbf::types::TbfFooterV2CredentialsType;
+
+/// What a AppCredentialsChecker decided a particular applications credential
+/// indicates about the runnability of an application binary.
+#[derive(Debug)]
+pub enum CheckResult {
+    /// Accept the credential and run the binary.
+    Accept,
+    /// Go to the next credential or in the case of the last one fall
+    /// back to the default policy.
+    Pass,
+    /// Reject the credential and do not run the binary.
+    Reject,
+}
+
+/// Receives callbacks on whether a credential was accepted or not.
+pub trait Client<'a> {
+    fn check_done(
+        &self,
+        result: Result<CheckResult, ErrorCode>,
+        credentials: TbfFooterV2Credentials,
+        binary: &'a [u8],
+    );
+}
+
+/// Implements a Credentials Checking Policy.
+pub trait AppCredentialsChecker<'a> {
+    fn set_client(&self, client: &'a dyn Client<'a>);
+    fn require_credentials(&self) -> bool;
+
+    fn check_credentials(
+        &self,
+        credentials: TbfFooterV2Credentials,
+        binary: &'a [u8],
+    ) -> Result<(), (ErrorCode, TbfFooterV2Credentials, &'a [u8])>;
+}
+
+/// Whether two processes have the same Application Identifier; two
+/// processes with the same Application Identifier cannot run concurrently.
+pub trait AppUniqueness {
+    /// Returns whether `process_a` and `process_b` have a different identifier,
+    /// and so can run concurrently. If this returns `false`, the kernel
+    /// will not run `process_a` and `process_b` at the same time.
+    fn different_identifier(&self, process_a: &dyn Process, process_b: &dyn Process) -> bool;
+
+    /// Return whether there is a currently running process that has
+    /// the same application identifier as `process`. This means that
+    /// if `process` is currently running, `has_unique_identifier`
+    /// returns false.
+    fn has_unique_identifier(
```

Comment:
It's not clear if the appid uniqueness property only applies if a verifier is in use, or always. It seems like the TRD indicates it is always in effect (and when uniqueness is not desired then the appid assignment should just use unique ids). If so, it seems like we should move this out of this trait and into the kernel proper. This is for two reasons, one it doesn't seem like a board needs to be able to adapt this, and two its implementation is tied to internal kernel details and how processes are stored, which we might not want to expose in this trait.

Reply:
The intention is that the uniqueness policy applies always. I think the way to formulate this is that if a process has a `short_id()` of `None`, this implies a unique value.

Reply:
Another question: does a non-unique process (aka one where this function returned false) have failed credentials? That is, would it be in `State::CredentialsFailed`?

Also, I'm still not clear on why this needs to be a customizable option as part of the Verifier.

Reply:
Is it valid for a board author to override `has_unique_identifier()` to do something such as always return `true`? It seems like `has_unique_identifier()` isn't subjective, either the identifiers are unique or not. However, the platform can choose to set the IDs such that they will always be unique (such as by using a counter or some other convention.

In general, if `AppUniqueness` always applies then it seems like it should be separate from `AppVerifier`.